### PR TITLE
#4 test integration

### PR DESCRIPTION
**Pull request by XioZ**
- Implemented `Mino` and `RPC` interfaces to make calls/streams
- Closing #4

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -1,67 +1,117 @@
 module go.dedis.ch/dela
 
-go 1.19
+go 1.21
```
Comment (ineiti) [inline go.mod]: Do you need a new version for libp2p? If not, revert these changes and stay with 1.19 - changing the go.mod of a big project is not an easy thing to do. E.g., 1.22 breaks some of the tests...

File: mino/minows/address.go [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+// String implements fmt.Stringer.
+func (a address) String() string {
+	return fmt.Sprintf("%s%s%s", a.location, protocolP2P, a.identity)
+}
+
+// ConnectionType implements mino.Address
+func (a address) ConnectionType() mino.AddressConnectionType {
+	// TODO implement
+	panic("not implemented")
```
Comment (ineiti) [inline mino/minows/address.go]: This looks easy to me to implement - is there something I'm missing?

File: mino/minows/address_test.go [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+	"go.dedis.ch/dela/mino"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+const (
+	AddrAllInterface = "/ip4/0.0.0.0/tcp/80"
+	AddrLocalhost    = "/ip4/127.0.0.1/tcp/80"
+	AddrHostname     = "/dns4/example.com/tcp/80"
```
Comment (ineiti) [inline mino/minows/address_test.go]: Please prepend all of these with `p2p://`, so it's `p2p://ipv4/0.0.0.0/tcp/80`. This will make it more similar to an URL.

File: mino/minows/address_test.go [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+		},
+		"missing location": {
+			in: []byte("/p2p/QmaD31nEzFGwD8dK96UFWHtTYTqYJgHLMYSFz4W4Hm2WCU"),
+		},
+		"missing identity": {
+			in: []byte(AddrLocalhost),
+		},
+	}
+
+	t.Parallel()
```
Comment (ineiti) [inline mino/minows/address_test.go]: From what I understand reading https://pkg.go.dev/testing#T.Parallel, this is not needed. Only the one on line 201.

File: mino/minows/integration_test.go [hunk 5]
```diff
@@ -0,0 +1,58 @@
+package minows
+
+//
+// import (
+// 	"context"
+// 	"testing"
+// 	"time"
+//
+// 	"github.com/stretchr/testify/require"
```
Comment (ineiti) [inline mino/minows/integration_test.go]: Remove commented out code.

File: mino/minows/mod.go [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+// TODO unit tests
+type minows struct {
+	myAddr   address
+	segments []string
+	host     host.Host
+	rpcs     map[string]rpc
+	context  serde.Context
+}
+
+// NewMinows
```
Comment (ineiti) [inline mino/minows/mod.go]: Add a short one-sentence description.

File: mino/minows/rpc.go [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+	return responses, nil
+}
+
+// Stream
+// - context defines when the protocol is done,
+// and it should therefore always be canceled at some point. (DELA Doc)
+// - When it's done, all the connections are shut down and the resources are
+// cleaned up (DELA Doc)
+// - orchestrator of a protocol will contact one of the participants which
+// will be the root for the routing algorithm (i.e. gateway?).
```
Comment (ineiti) [inline mino/minows/rpc.go]:
```suggestion
// will be the root for the routing algorithm (the orchestrator).
```

File: mino/minows/rpc.go [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+// It will then relay the messages according to the routing algorithm and
+// create relays to other peers when necessary (DELA Doc but ignored)
```
Comment (ineiti) [inline mino/minows/rpc.go]:
```suggestion
// It will then relay the messages according to the routing algorithm, which is
// a direct connection between every pair of nodes in our case.
```

**Review by ineiti — CHANGES_REQUESTED** [review]
Starts looking good. Needs some changes.

Also there are a lot of TODOs in the code - do you have any idea if all of them are necessary for the final implementation, or some of them are optional?

File: mino/minows/address.go [hunk 9]
```diff
@@ -86,7 +86,7 @@ func (f addressFactory) FromText(text []byte) mino.Address {
 	}
 	addr, err := newAddress(location, identity)
 	if err != nil {
-		dela.Logger.Err(xerrors.Errorf("could not create address: %v", err))
+		dela.Logger.Err(xerrors.Errorf("could not create address: %w", err))
```
Comment (ineiti) [inline mino/minows/address.go]: Please revert. The standard error-passing mechanism in dela is using `%v`.

File: mino/minows/address_test.go [hunk 10]
```diff
@@ -9,65 +9,78 @@ import (
...
+			args: args{addrLocalhost, pid2},
 		},
 		"hostname": {
-			location: mustCreateMultiaddress(AddrHostname),
-			identity: mustCreatePeerID(
-				PID1),
+			args: args{addrHostname, pid1},
 		},
 	}
+	t.Parallel() // run this test function args parallel to other test functions
```
Comment (ineiti) [inline mino/minows/address_test.go]: Here and elsewhere: the dela guid asks comments to be on their own line. So this should be:
```suggestion
    // run this test function args parallel to other test functions
	t.Parallel() 
```

And commenting obvious things should be avoided, like this line.

File: mino/minows/address_test.go [hunk 11]
```diff
@@ -9,65 +9,78 @@ import (
...
+	}{
+		"missing location": {
+			location: nil,
+			identity: mustCreatePeerID(t, pid1),
+		},
+		"missing identity": {
+			location: mustCreateMultiaddress(t, addrAllInterface),
+			identity: "",
+		}}
 	t.Parallel() // run this test function in parallel to other test functions
```
Comment (ineiti) [inline mino/minows/address_test.go]: Aren't all tests run in parallel anyway? Except the ones where you call `t.Run`, where `t.Parallel` makes sense.

File: mino/minows/address_test.go [hunk 12]
```diff
@@ -9,65 +9,78 @@ import (
...
-	PID2 = "QmVt9t5Tk2uEoA4CDbKNCVxqrut8UXmWHXvFZ8wFZ3ghhv"
-)
-
-func TestNewAddress(t *testing.T) {
+func Test_newAddress(t *testing.T) {
+	const pid1 = "QmaD31nEzFGwD8dK96UFWHtTYTqYJgHLMYSFz4W4Hm2WCU"
+	const pid2 = "QmVt9t5Tk2uEoA4CDbKNCVxqrut8UXmWHXvFZ8wFZ3ghhv"
+	const addrAllInterface = "/ip4/0.0.0.0/tcp/80"
+	const addrLocalhost = "/ip4/127.0.0.1/tcp/80"
+	const addrHostname = "/dns4/example.com/tcp/80"
```
Comment (ineiti) [inline mino/minows/address_test.go:17]: Please add some protocol identifier like `p2pws://ip4/0.0.0.0/tcp/80`

File: mino/minows/integration_test.go [hunk 13]
```diff
@@ -45,7 +45,7 @@ package minows
 // }
 //
 // func newMinowsAndRPC() (mino.Mino, mino.RPC) {
-// 	m, err := NewMinows("ip4/127.0.0.1/tcp/9998/ws", "ip4/127.0.0.1/tcp/9998/ws",
+// 	m, err := newMinows("ip4/127.0.0.1/tcp/9998/ws", "ip4/127.0.0.1/tcp/9998/ws",
```
Comment (ineiti) [inline mino/minows/integration_test.go]: For the whole file: please don't commit commented out code. Remove the commented code from the commit.

File: mino/minows/mod.go [hunk 14]
```diff
@@ -17,40 +20,43 @@ import (
...
+	rpcs      map[string]any
 }
 
-// NewMinows
-// listen: local listening address in multiaddress format, e.g. /ip4/0.0.0.0/tcp/80
-// public: public dial-able address in multiaddress format, e.g. /dns4/p2p-1.c4dt.dela.org/tcp/443/wss
-// identity: Peer ID (multi-hash of public key generated by libp2p)
-func NewMinows(listen, public ma.Multiaddr, privKey crypto.PrivKey) (*minows, error) {
-	id, err := peer.IDFromPrivateKey(privKey)
+// newMinows
```
Comment (ineiti) [inline mino/minows/mod.go]: There should be a phrase at the beginning, e.g.,

```
// newMinows starts a new mino service using libp2p to connect over websockets.
```

File: mino/minows/mod.go [hunk 15]
```diff
@@ -17,40 +20,43 @@ import (
...
+// listen: local listening address in multiaddress format,
+// e.g. /ip4/0.0.0.0/tcp/80
+// public: public dial-able address in multiaddress format,
+// e.g. /dns4/p2p-1.c4dt.dela.org/tcp/443/wss
+func newMinows(listen, public ma.Multiaddr, secret crypto.PrivKey) (*minows,
+	error) {
+	id, err := peer.IDFromPrivateKey(secret)
 	if err != nil {
-		return nil, xerrors.Errorf("could not get Peer ID: %v", err)
+		return nil, xerrors.Errorf("could not get Peer ID: %w", err)
```
Comment (ineiti) [inline mino/minows/mod.go]: Here and elsewhere: please use `%v`

File: mino/minows/mod.go [hunk 16]
```diff
@@ -87,18 +97,74 @@ func (m *minows) CreateRPC(name string, h mino.Handler, f serde.Factory) (mino.R
...
+	}
+	// TODO need to be thread safe?
+	m.rpcs[name] = nil
+	// start listening
+	m.host.SetStreamHandler(protocol.ID(uri+PostfixCall),
+		r.createCallHandler(h))
+	m.host.SetStreamHandler(protocol.ID(uri+PostfixStream),
+		r.createStreamHandler(h))
+	// TODO when to return pointer vs struct?
+	return r, nil
```
Comment (ineiti) [inline mino/minows/mod.go:122]: https://bryanftan.medium.com/accept-interfaces-return-structs-in-go-d4cab29a301b

But I'm not sure how dela handles this in other modules.

**Review by ineiti — CHANGES_REQUESTED** [review]
Looks good - there are some changes requested - please change them, if possible before our meeting.

**Comment by ineiti** [comment]
- Needs to be rebased on the latest `main`
- Please include my comments

File: go.mod (hunk 1 above)
Comment (XioZ) [reply go.mod]: didn't realize the go version got updated automatically when i ran `go get libp2p`. reverted `go.mod` file and using go-libp2p v0.29 now (instead of 0.33) to be compatible with go v1.19. tests still pass

File: mino/minows/address.go (hunk 2 above)
Comment (XioZ) [reply mino/minows/address.go]: adding a new `AddressConnectionType` const previously broke some tests in grpc, so left it last to implement. returning `ACTws` now but this method is not used.

File: mino/minows/address_test.go (hunk 3 above)
Comment (XioZ) [reply mino/minows/address_test.go]: multiaddress requires a certain format. [this doc](https://docs.libp2p.io/concepts/fundamentals/addressing/) contains more details.

File: mino/minows/address_test.go (hunk 4 above)
Comment (XioZ) [reply mino/minows/address_test.go]: first call parallelizes the top-level test function to other test functions. second parallelizes the subtest to other subtests in the table test. [this doc](https://go.dev/wiki/TableDrivenTests#parallel-testing) contains more details.

File: mino/minows/integration_test.go (hunk 5 above)
Comment (XioZ) [reply mino/minows/integration_test.go]: removed file

File: mino/minows/mod.go (hunk 6 above)
Comment (XioZ) [reply mino/minows/mod.go]: auto generated doc last. please check updated comments

File: mino/minows/address.go (hunk 9 above)
Comment (XioZ) [reply mino/minows/address.go]: changed, though i'm not sure how this makes error assertion `errors.Is()` still possible at the caller.

File: mino/minows/address_test.go (hunk 10 above)
Comment (XioZ) [reply mino/minows/address_test.go]: moved to own lines. maybe it's not obvious at first glance, since the two parallel calls are parallelizing diff things?

File: mino/minows/address_test.go (hunk 11 above)
Comment (XioZ) [reply mino/minows/address_test.go]: from [this doc](https://go.dev/wiki/TableDrivenTests#parallel-testing) it looks like by default `t.Run()` runs test functions sequentially? correct me if i misunderstood.

File: mino/minows/address_test.go (hunk 12 above)
Comment (XioZ) [reply mino/minows/address_test.go:17]: `ip4` and `tcp` are the predefined protocol identifiers according to [this doc ](https://pkg.go.dev/github.com/multiformats/go-multiaddr#readme-example)

File: mino/minows/integration_test.go (hunk 13 above)
Comment (XioZ) [reply mino/minows/integration_test.go]: removed file

File: mino/minows/mod.go (hunk 15 above)
Comment (XioZ) [reply mino/minows/mod.go]: changed

File: mino/minows/mod.go (hunk 14 above)
Comment (XioZ) [reply mino/minows/mod.go]: generated doc now

File: mino/minows/mod.go (hunk 16 above)
Comment (XioZ) [reply mino/minows/mod.go:122]: returns pointer now

**Comment by XioZ** [comment]
just to clarify, the todos and comments were mostly for my own reference during development, and the pr was marked as draft to indicate not ready for review. i've resolved the todos now (only 1 left on `MaxMessageSize`) and generated doc so now it should be in shape for review.

File: mino/minows/rpc.go [hunk 17]
```diff
@@ -0,0 +1,318 @@
...
+	"io"
+	"sync"
+
+	"github.com/libp2p/go-libp2p/core/protocol"
+	"go.dedis.ch/dela/mino"
+	"go.dedis.ch/dela/serde"
+	"golang.org/x/xerrors"
+)
+
+const MaxMessageSize = 1e9 // TODO verify
```
Comment (XioZ) [inline mino/minows/rpc.go]: copied this from grpc as the buffer size to use when reading from byte stream. please confirm it's ok?

File: mino/minows/address_test.go (hunk 12 above)
Comment (ineiti) [reply mino/minows/address_test.go:17]: I'm more concerned with mino-identifiers. Like the `grpc://` and `grpc://` ones used for minogrpc.

On second thought I'm not sure it makes that much sense, as one dela-cli can only handle one mino-implementation anyway. Let's leave it like that.

File: mino/minows/address.go (hunk 9 above)
Comment (ineiti) [reply mino/minows/address.go]: Not sure, either. But sometimes it's more important to use common style than to support a feature nobody uses.

If you think that `errors.Is()` is not possible, open an issue on the dedis/dela repository, and express your concerns.

File: mino/minows/address_test.go (hunk 10 above)
Comment (ineiti) [reply mino/minows/address_test.go]: Yes, true. I was confused by the fact that tests from multiple packages are run in parallel, but the tests within one package are run sequentially:

https://bryce.is/writing/code/go-test-and-parallelism

(Which also explains the PAUSE we saw in one of the test runs)

Again, for the unity of the code, I would not run the tests in parallel here. Or only if you have a really good reason, like running the tests 10'000 times or so.

I try to avoid surprises like these, for debugging's sake. As all the other tests in `dela` are sequential, please make these tests also sequential.

File: mino/minows/address_test.go (hunk 11 above)
Comment (ineiti) [reply mino/minows/address_test.go]: You're correct. It's tests from different packages which are run in parallel. See my comment above.

**Comment by ineiti** [comment]
> just to clarify, the todos and comments were mostly for my own reference during development, and the pr was marked as draft to indicate not ready for review. i've resolved the todos now (only 1 left on `MaxMessageSize`) and generated doc so now it should be in shape for review.

OK, point taken. You did request a review from me, though. This happens if you choose one of the collaborators as `Reviewer`. So you should:

- start in draft
- work on it
- put it in 'ready for merge'
- request a review - if possible within the first half of the week

File: mino/minows/address_test.go (hunk 4 above)
Comment (ineiti) [reply mino/minows/address_test.go]: See my other comments - why add `Parallel` at all here? It's nothing taking a long time, and no other package in dela uses `Parallel`. So the confusion / usefulness ratio is too high in my opinion.

Remove `Parallel` for your tests, unless you have a good reason to have it here.

File: mino/minows/mod_test.go [hunk 18]
```diff
@@ -0,0 +1,212 @@
...
+	const publicWS = "/ip4/127.0.0.1/tcp/80/ws"
+	const publicWSS = "/ip4/127.0.0.1/tcp/443/wss"
+	type args struct {
+		listen string
+		public string
+	}
+	var tests = map[string]struct {
+		args args
+	}{
+		// 'public' only uses localhost a for local testing
```
Comment (ineiti) [inline mino/minows/mod_test.go]: What do you mean by that?

File: mino/minows/rpc.go [hunk 19]
```diff
@@ -0,0 +1,320 @@
...
+// Stream is blocking and returns only after all communications are
+// established.
+// Returns an error if no players, any player address is of the wrong type,
+// or communication to any player fails to establish.
+// Otherwise, returns a stream session that can be used to send and receive
+// messages.
+// Note:
+// - 'ctx' defines when the stream session ends,
+// so should always be canceled at some point (e.g. completed task or errored).
+// - When it's done, all the connections are shut down and resources freed.
```
Comment (ineiti) [inline mino/minows/rpc.go]: This is not clear to me. `ctx` MUST cancel to close down?
When is `done`? When `ctx` is canceled?

File: mino/minows/rpc.go [hunk 20]
```diff
@@ -0,0 +1,320 @@
...
+func (r rpc) Stream(ctx context.Context, players mino.Players) (mino.Sender, mino.Receiver, error) {
+	if players == nil || players.Len() == 0 {
+		return nil, nil, xerrors.New("no players to stream")
+	}
+	// quit unless all player addresses are valid
```
Comment (ineiti) [inline mino/minows/rpc.go]: Avoid negations in comments, it takes more time to read. Better:
```
// quit if any address is invalid
```

or

```
// Ensure all addresses are valid
```

or add this to the comment at the top of the method. If somebody reads these lines, they will already understand what's going on. But somebody might only read the documentation.

File: mino/minows/mod.go [hunk 21]
```diff
@@ -0,0 +1,125 @@
...
+		myAddr:   m.myAddr,
+		segments: append(m.segments, segment),
+		host:     m.host,
+		rpcs:     make(map[string]any), // don't copy existing rpcs
+		context:  m.context,
+	}
+}
+
+func (m *minows) CreateRPC(name string, h mino.Handler, f serde.Factory) (mino.RPC, error) {
+	pattern := regexp.MustCompile("^[a-zA-Z0-9]+$")
```
Comment (ineiti) [inline mino/minows/mod.go]: You could optimize it a bit by adding this into a package-var, like here:

https://github.com/dedis/dela/blob/master/mino/minogrpc/mod.go#L417

File: mino/minows/mod.go [hunk 22]
```diff
@@ -0,0 +1,125 @@
...
+	// validate namespace once
+	if len(m.rpcs) == 0 {
```
Comment (ineiti) [inline mino/minows/mod.go:90]: I wonder whether we should put that check in the `minows.WithSegment` method. The only downside is that `WithSegment` doesn't return an error, and you'd have to panic.

But having an invalid segment name would be a valid reason for me to panic.

File: mino/minows/mod.go [hunk 23]
```diff
@@ -0,0 +1,125 @@
...
+		mino:    m,
+		factory: f,
+		context: m.context,
+	}
+	// register handlers
+	m.host.SetStreamHandler(protocol.ID(uri+PostfixCall),
+		r.createCallHandler(h))
+	m.host.SetStreamHandler(protocol.ID(uri+PostfixStream),
+		r.createStreamHandler(h))
+	m.rpcs[name] = nil
```
Comment (ineiti) [inline mino/minows/mod.go:120]: Why do you delete the rpc entry here? Now you can create twice the same RPC name, no?

**Review by ineiti — COMMENTED** [review]
- Please add some empty lines in your methods, according to https://dedis.github.io/dela/#/guideline
- avoid useless comments, like `form uri` or `create rpc` in the `minows.CreateRPC` method. This only catches the attention of the reader and doesn't help.
- still some comments on the same line as the code, please put on their own line
- comments should be a complete sentence (subject, verb), start with a capital letter, and end with a `.`
- rebase on the latest main
- please respect my week-end and don't ask me to review on Friday afternoons!

**Comment by XioZ** [comment]
> > just to clarify, the todos and comments were mostly for my own reference during development, and the pr was marked as draft to indicate not ready for review. i've resolved the todos now (only 1 left on `MaxMessageSize`) and generated doc so now it should be in shape for review.
> 
> OK, point taken. You did request a review from me, though. This happens if you choose one of the collaborators as `Reviewer`. So you should:
> 
> * start in draft
> * work on it
> * put it in 'ready for merge'
> * request a review - if possible within the first half of the week

i think there's a miscommunication here. 2 meetings ago, you asked me to open a new PR whenever i start working on an issue and add you as a reviewer so you see my progress. so i did that with the draft PR and pushed my commits occasionally. i marked this PR ready for review today after seeing your comments. i don't mean to pressure you to review it right away, sorry. please do it whenever is convenient and i'll keep working on the next issue meanwhile. i'll address your comments as they come along. have a good weekend.

File: mino/minows/rpc.go (hunk 19 above)
Comment (XioZ) [reply mino/minows/rpc.go]: a [context is done](https://pkg.go.dev/context#Context) when cancelled or times out (depending on how it's created), or its parent context is done. i copied these 2 bullet points from [dela doc](https://dedis.github.io/dela/#/mino?id=stream-bidirectional-stream-based-protocol). please let me know how it can be rephrased for clarity

File: mino/minows/mod.go (hunk 23 above)
Comment (XioZ) [reply mino/minows/mod.go:120]: thereafter, `_, found := m.rpcs[name]` would return `found` as true. please see `Test_minows_CreateRPC_AlreadyExists` for verified behavior.

File: mino/minows/mod.go (hunk 21 above)
Comment (XioZ) [reply mino/minows/mod.go]: changed

File: mino/minows/rpc.go [hunk 24]
```diff
@@ -0,0 +1,332 @@
...
+// Stream is blocking and returns only after all communications are
+// established.
+// Returns an error if no players, any player address is of the wrong type,
+// or communication to any player fails to establish.
+// Otherwise, returns a stream session that can be used to send and receive
+// messages.
+// Note:
+// - 'ctx' defines when the stream session ends,
+// so should always be canceled at some point (e.g. completed task or errored).
+// - When it's done, all the connections are shut down and resources freed.
```
Comment (ineiti) [inline mino/minows/rpc.go]:
```suggestion
// Note: to end the session, the 'ctx' needs to get into 'done' state by cancelling, raising an error, or completing the task. When the session ends, all the connections are shut down and resources freed.
```

**Comment by XioZ** [comment]
Addressed PR comments. Uses a JSON encoder/decoder to write to/read from libp2p streams (instead of manually allocating a byte buffer each time) to address memory concerns.

**Review by ineiti — APPROVED** [review]
